### PR TITLE
SUPESC-334: SOFORT Payment is authorizing successfully.

### PR DESCRIPTION
**Pull request by kkicheglovspryker**
- Developer(s): @kkicheglovspryker 

- Ticket: https://spryker.atlassian.net/browse/SUPESC-334

- Academy PR: -

- Release Group: https://release.spryker.com/release-groups/view/3435

- merge: merge

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   Adyen                 | minor                 |                       |

-----------------------------------------

##### Change log

Improvements

- Adjusted `AdyenFacade::handleOnlineTransferResponseFromAdyen()` and `AdyenFacade::handleCreditCard3dResponseFromAdyen()` therefore the order status does not change if it is already set by the Adyen webhook instead of overriding the status when processing the Adyen response.

### CODE REVIEW

File: tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php [hunk 1]
```diff
@@ -228,6 +228,28 @@ public function testHandleOnlineTransferResponseFromAdyen(): void
         }
     }
 
+    /**
+     * @return void
+     */
+    public function testHandleOnlineTransferResponseFromAdyenAfterNotification(): void
+    {
+        $facade = $this->createFacade();
```
Comment (AsonUnique) [inline tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php:236]: Please, add `// Arrange // Act // Assert` comments. 
Check for example `Bundles/StockAddress/tests/SprykerTest/Zed/StockAddress/Business/StockAddressFacadeTest.php`

**Review by AsonUnique — APPROVED** [review]

**Review by romanhavrylko — APPROVED** [review]

**Review by volodymyr-nakvasiuk — APPROVED** [review]

File: tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php [hunk 2]
```diff
@@ -228,6 +228,34 @@ public function testHandleOnlineTransferResponseFromAdyen(): void
...
+        $redirectResponseTransfer = $this->createRedirectResponseTransfer($orderTransfer);
+
+        //Act
+        $result = $facade->handleOnlineTransferResponseFromAdyen($redirectResponseTransfer);
+
+        // Assert
+        $this->assertTrue($result->getIsSuccess());
+
+        foreach ($this->getSpySalesOrderItems($orderTransfer) as $item) {
+            // Assert
```
Comment (abitskil) [inline tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php:251]: * Remove this `//Assert`.

File: src/SprykerEco/Zed/Adyen/Business/Handler/Redirect/OnlineTransferRedirectHandler.php [hunk 3]
```diff
@@ -119,7 +119,13 @@ protected function processPaymentDetailsResponse(PaymentAdyenTransfer $paymentAd
         $paymentAdyenTransfer->setPspReference($responseTransfer->getPaymentDetailsResponse()->getPspReference());
         $paymentAdyenOrderItems = $this->reader->getAllPaymentAdyenOrderItemsByIdSalesOrder($paymentAdyenTransfer->getFkSalesOrder());
 
-        $this->writer->updatePaymentEntities($this->getOmsStatus(), $paymentAdyenOrderItems, $paymentAdyenTransfer);
+        $defaultOmsStatus = $this->getOmsStatus();
```
Comment (abitskil) [inline src/SprykerEco/Zed/Adyen/Business/Handler/Redirect/OnlineTransferRedirectHandler.php:122]: * Extract to the method `resolveOmsStatus()`.

**Review by abitskil — CHANGES_REQUESTED** [review]
Rewrite the change logs in the context of the public API:
* Adjusted `AdyenFacade::handleOnlineTransferResponseFromAdyen(),AdyenFacade::handleCreditCard3dResponseFromAdyen()`.
* Use the documentation https://spryker.atlassian.net/wiki/spaces/CORE/pages/523730976/Release+Notes+and+Changelogs.

Fix the code sniffer error:
* `Every API public method must also contain the {@inheritDoc} tag in docblock.`.

Fix the `phpstan` failing.

File: src/SprykerEco/Yves/Adyen/Plugin/Payment/CreditCardPaymentMapperPlugin.php [hunk 4]
```diff
@@ -15,6 +15,8 @@
 class CreditCardPaymentMapperPlugin implements AdyenPaymentMapperPluginInterface
 {
     /**
+     * {@inheritDoc}
+     *
```
Comment (abitskil) [inline src/SprykerEco/Yves/Adyen/Plugin/Payment/CreditCardPaymentMapperPlugin.php:19]: * Add an `@api` mark to a method.
* Do the same for a similar places.

Comment (abitskil) [reply src/SprykerEco/Yves/Adyen/Plugin/Payment/CreditCardPaymentMapperPlugin.php:19]: Enable GitHub actions by a copy them from a repository: 
* https://github.com/spryker-eco/avalara-tax-product-option/pull/3/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03f

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                  composer --version
+                  composer install
+            - name: Setup PHP
+              uses: shivammathur/setup-php@v2
+              with:
+                  php-version: ${{ matrix.php }}
+                  extensions: mbstring, intl, pdo_mysql
+                  tools: composer:v2
+
+#       suite-tests:
```
Comment (abitskil) [inline .github/workflows/ci.yml:57]: * Remove commented part.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,67 @@
+name: CI
+
+on: [
+        push,
+        workflow_dispatch
+]
+
+jobs:
```
Comment (abitskil) [inline .github/workflows/ci.yml:8]: * Copy the content from https://github.com/spryker-eco/loggly/blob/master/.github/workflows/ci.yml.

**Review by dmytro-sokolovskyi — APPROVED** [review]

File: src/SprykerEco/Service/Adyen/Dependency/Service/AdyenToUtilTextServiceInterface.php [hunk 7]
```diff
@@ -15,5 +15,5 @@ interface AdyenToUtilTextServiceInterface
      *
      * @return string
      */
-    public function hashValue($value, $algorithm);
+    public function hashValue($value, string $algorithm): string;
```
Comment (abitskil) [inline src/SprykerEco/Service/Adyen/Dependency/Service/AdyenToUtilTextServiceInterface.php:18]: The bridge method must have the same signature as the method that it uses.

File: src/SprykerEco/Yves/Adyen/Dependency/Service/AdyenToUtilEncodingServiceInterface.php [hunk 8]
```diff
@@ -16,7 +16,7 @@ interface AdyenToUtilEncodingServiceInterface
      *
      * @return string|null
      */
-    public function encodeJson($value, $options = null, $depth = null);
+    public function encodeJson(array $value, ?int $options = null, ?int $depth = null): ?string;
```
Comment (abitskil) [inline src/SprykerEco/Yves/Adyen/Dependency/Service/AdyenToUtilEncodingServiceInterface.php:19]: Same issue as above.

File: src/SprykerEco/Yves/Adyen/Plugin/Provider/AdyenControllerProvider.php [hunk 9]
```diff
@@ -24,7 +24,7 @@ class AdyenControllerProvider extends AbstractYvesControllerProvider
      *
      * @return void
      */
-    protected function defineControllers(Application $app)
+    protected function defineControllers(Application $app): void
```
Comment (abitskil) [inline src/SprykerEco/Yves/Adyen/Plugin/Provider/AdyenControllerProvider.php:27]: Undo the change.

File: src/SprykerEco/Zed/Adyen/Dependency/Facade/AdyenToCalculationFacadeInterface.php [hunk 10]
```diff
@@ -16,5 +16,5 @@ interface AdyenToCalculationFacadeInterface
      *
      * @return \Generated\Shared\Transfer\OrderTransfer
      */
-    public function recalculateOrder(OrderTransfer $orderTransfer);
+    public function recalculateOrder(OrderTransfer $orderTransfer): OrderTransfer;
```
Comment (abitskil) [inline src/SprykerEco/Zed/Adyen/Dependency/Facade/AdyenToCalculationFacadeInterface.php:19]: Same issues as above.

File: src/SprykerEco/Client/Adyen/Dependency/Client/AdyenToZedRequestClientInterface.php [hunk 11]
```diff
@@ -18,5 +18,5 @@ interface AdyenToZedRequestClientInterface
      *
      * @return \Spryker\Shared\Kernel\Transfer\TransferInterface
      */
-    public function call($url, TransferInterface $object, $requestOptions = null);
+    public function call(string $url, TransferInterface $object, $requestOptions = null): TransferInterface;
```
Comment (abitskil) [inline src/SprykerEco/Client/Adyen/Dependency/Client/AdyenToZedRequestClientInterface.php:21]: Checke the original method.

File: src/SprykerEco/Service/Adyen/Dependency/Service/AdyenToUtilTextServiceInterface.php [hunk 12]
```diff
@@ -15,5 +15,6 @@ interface AdyenToUtilTextServiceInterface
      *
      * @return string
      */
-    public function hashValue($value, $algorithm);
+    //phpcs:ignore
+    public function hashValue($value, $algorithm): string;
```
Comment (abitskil) [inline src/SprykerEco/Service/Adyen/Dependency/Service/AdyenToUtilTextServiceInterface.php:19]: Same.

File: src/SprykerEco/Yves/Adyen/Plugin/Payment/CreditCardPaymentMapperPlugin.php [hunk 13]
```diff
@@ -15,6 +15,13 @@
 class CreditCardPaymentMapperPlugin implements AdyenPaymentMapperPluginInterface
 {
     /**
+     * Specification:
+     *  - Sets CreditCard payment data to Quote.
+     *
+     * {@inheritDoc}
```
Comment (abitskil) [inline src/SprykerEco/Yves/Adyen/Plugin/Payment/CreditCardPaymentMapperPlugin.php:21]: * It should be higher than specification.

File: src/SprykerEco/Yves/Adyen/Plugin/Payment/CreditCardPaymentMapperPlugin.php [hunk 14]
```diff
@@ -15,6 +15,13 @@
 class CreditCardPaymentMapperPlugin implements AdyenPaymentMapperPluginInterface
 {
     /**
+     * Specification:
+     *  - Sets CreditCard payment data to Quote.
```
Comment (abitskil) [inline src/SprykerEco/Yves/Adyen/Plugin/Payment/CreditCardPaymentMapperPlugin.php:19]: * Wrap names in backticks.
* Check other places.

File: src/SprykerEco/Yves/Adyen/Plugin/StepEngine/AdyenAliPaySubFormPlugin.php [hunk 15]
```diff
@@ -26,6 +33,13 @@ public function createSubForm(): SubFormInterface
     }
 
     /**
+     * Specification:
+     * - Creates AliPay subform DataProvider.
```
Comment (abitskil) [inline src/SprykerEco/Yves/Adyen/Plugin/StepEngine/AdyenAliPaySubFormPlugin.php:37]: data provider.

File: src/SprykerEco/Zed/Adyen/Dependency/Facade/AdyenToSalesFacadeInterface.php [hunk 16]
```diff
@@ -14,5 +14,6 @@ interface AdyenToSalesFacadeInterface
      *
      * @return \Generated\Shared\Transfer\OrderTransfer
      */
-    public function getOrderByIdSalesOrder($idSalesOrder);
+    //phpcs:ignore
+    public function getOrderByIdSalesOrder(int $idSalesOrder);
```
Comment (abitskil) [inline src/SprykerEco/Zed/Adyen/Dependency/Facade/AdyenToSalesFacadeInterface.php:18]: Remove the type hint.

File: src/SprykerEco/Client/Adyen/Dependency/Client/AdyenToZedRequestClientInterface.php [hunk 17]
```diff
@@ -18,5 +18,6 @@ interface AdyenToZedRequestClientInterface
      *
      * @return \Spryker\Shared\Kernel\Transfer\TransferInterface
      */
-    public function call($url, TransferInterface $object, $requestOptions = null);
+    //phpcs:ignore
+    public function call($url, TransferInterface $object, $requestOptions = null): TransferInterface;
```
Comment (abitskil) [inline src/SprykerEco/Client/Adyen/Dependency/Client/AdyenToZedRequestClientInterface.php:22]: * Remove the return type hint.

File: src/SprykerEco/Yves/Adyen/Dependency/Client/AdyenToQuoteClientInterface.php [hunk 18]
```diff
@@ -14,12 +14,12 @@ interface AdyenToQuoteClientInterface
     /**
      * @return \Generated\Shared\Transfer\QuoteTransfer
      */
-    public function getQuote();
+    public function getQuote(): QuoteTransfer;
```
Comment (abitskil) [inline src/SprykerEco/Yves/Adyen/Dependency/Client/AdyenToQuoteClientInterface.php:17]: * Remove the return type hint.

File: src/SprykerEco/Yves/Adyen/Dependency/Client/AdyenToQuoteClientInterface.php [hunk 19]
```diff
@@ -14,12 +14,12 @@ interface AdyenToQuoteClientInterface
...
 
     /**
      * @param \Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
      *
      * @return void
      */
-    public function setQuote(QuoteTransfer $quoteTransfer);
+    public function setQuote(QuoteTransfer $quoteTransfer): void;
```
Comment (abitskil) [inline src/SprykerEco/Yves/Adyen/Dependency/Client/AdyenToQuoteClientInterface.php:24]: * Remove the return type hint.

File: src/SprykerEco/Yves/Adyen/Dependency/Service/AdyenToUtilEncodingServiceInterface.php [hunk 20]
```diff
@@ -16,7 +16,8 @@ interface AdyenToUtilEncodingServiceInterface
      *
      * @return string|null
      */
-    public function encodeJson($value, $options = null, $depth = null);
+    //phpcs:ignore
+    public function encodeJson($value, $options = null, $depth = null): ?string;
```
Comment (abitskil) [inline src/SprykerEco/Yves/Adyen/Dependency/Service/AdyenToUtilEncodingServiceInterface.php:20]: * Remove the return type hint.

File: src/SprykerEco/Yves/Adyen/Plugin/Payment/KlarnaInvoicePaymentMapperPlugin.php [hunk 21]
```diff
@@ -18,6 +18,13 @@
 class KlarnaInvoicePaymentMapperPlugin implements AdyenPaymentMapperPluginInterface
 {
     /**
+     * {@inheritDoc}
+     *
+     * Specification:
```
Comment (abitskil) [inline src/SprykerEco/Yves/Adyen/Plugin/Payment/KlarnaInvoicePaymentMapperPlugin.php:23]: * The word `Specification` is not needed in this case.
* The specification should go right after `{@inheritDoc}`.
*` \Spryker\Service\ProductConfigurationStorage\Plugin\PriceProduct\ProductConfigurationPriceProductVolumeFilterPlugin`

File: src/SprykerEco/Yves/Adyen/Plugin/StepEngine/AdyenDirectDebitSubFormPlugin.php [hunk 22]
```diff
@@ -18,6 +18,13 @@
 class AdyenDirectDebitSubFormPlugin extends AbstractPlugin implements SubFormPluginInterface
 {
     /**
+     * {@inheritDoc}
+     * - Creates `DirectDebit` subform.
+     *
+     * {@inheritDoc}
```
Comment (abitskil) [inline src/SprykerEco/Yves/Adyen/Plugin/StepEngine/AdyenDirectDebitSubFormPlugin.php:24]: * Remove this line.

File: src/SprykerEco/Yves/Adyen/Plugin/StepEngine/AdyenDirectDebitSubFormPlugin.php [hunk 23]
```diff
@@ -26,6 +33,13 @@ public function createSubForm(): SubFormInterface
     }
 
     /**
+     * {@inheritDoc}
+     * - Creates `DirectDebit` subform data provider.
+     *
+     * {@inheritDoc}
```
Comment (abitskil) [inline src/SprykerEco/Yves/Adyen/Plugin/StepEngine/AdyenDirectDebitSubFormPlugin.php:39]: * Remove this line.

**Review by abitskil — APPROVED** [review]

**Pull request merged**